src/include/function/gds/gds_frontier.h

```diff
@@ -39,21 +86,52 @@ class RangeFrontierMorsel {
     offset_t nextOffset = INVALID_OFFSET;
 };
 
-/**
- * Base interface for algorithms that can be implemented in Pregel-like vertex-centric manner.
- * More specifically, this interface mimics Ligra's edgeUpdate function (though we call it
- * edgeCompute). Intended to be used when using the helper functions in GDSUtils. The name
- * FrontierCompute comes from vertex.compute() or edge.compute() functions of these systems.
- */
-class FrontierCompute {
+class FrontierMorselDispatcher {
+    static constexpr uint64_t MIN_FRONTIER_MORSEL_SIZE = 512;
+    // Note: MIN_NUMBER_OF_FRONTIER_MORSELS is the minimum number of morsels we aim to have but we
+    // can have fewer than this. See the beginFrontierComputeBetweenTables to see the actual
+    // morselSize computation for details.
+    static constexpr uint64_t MIN_NUMBER_OF_FRONTIER_MORSELS = 128;
+
 public:
-    virtual ~FrontierCompute() = default;
-    // Does any work that is needed while extending the (curNodeID, nbrID) edge. curNodeID is the
-    // nodeID that is in the current frontier and currently executing. Returns true if the neighbor
-    // should be put in the next frontier. So if the implementing class has access to the next
-    // frontier as a field, **do not** call setActive. Helper functions in GDSUtils will do that
-    // work.
-    virtual bool edgeCompute(nodeID_t curNodeID, nodeID_t nbrID) = 0;
+    explicit FrontierMorselDispatcher(uint64_t _maxThreadsForExec) {
+        maxThreadsForExec.store(_maxThreadsForExec);
+    }
+
+    void init(table_id_t _tableID, uint64_t _numOffsets) {
+        tableID.store(_tableID);
+        numOffsets.store(_numOffsets);
+        nextOffset.store(0u);
+        // Frontier size calculation: The ideal scenario is to have k^2 many morsels where k
+        // the number of maximum threads that could be working on this frontier. However if
+        // that is too small then we default to MIN_FRONTIER_MORSEL_SIZE.
+        auto idealMorselSize = numOffsets.load(std::memory_order_relaxed) /
+                               (std::max(MIN_NUMBER_OF_FRONTIER_MORSELS,
+                                   maxThreadsForExec.load(std::memory_order_relaxed) *
+                                       maxThreadsForExec.load(std::memory_order_relaxed)));
+        morselSize = std::max(MIN_FRONTIER_MORSEL_SIZE, idealMorselSize);
+    }
+
+    bool getNextRangeMorsel(FrontierMorsel& frontierMorsel) {
```

Comment:
Done. What's the convention for what gets moved to cpp and what stays in header files. I wonder if we should just have a simple rule that says everything should move to cpp (though I don't know if this has any consequences for optimizer inlining functions, e.g., that functions defined in cpp files are never inlined).

Reply:
one/or very few lines code can exist in header because it will always be inlined. 

Its true functions defined in cpp will never be inlined. In your case, the function body is heavy and won't be inlined anyways. Reducing header size is good for compilation time in general.